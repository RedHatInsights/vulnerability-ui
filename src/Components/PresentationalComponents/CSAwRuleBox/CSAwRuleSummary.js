/* eslint-disable react/no-danger */
import React from 'react';
import marked from 'marked';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import sanitizeHtml from 'sanitize-html';
import { Truncate } from '@redhat-cloud-services/frontend-components';
import { Stack, StackItem, TextContent } from '@patternfly/react-core';
import { handleCVELink } from '../../../Helpers/VulnerabilityHelper';
import { TRUNCATE_TEXT_THRESHOLD } from '../../../Helpers/constants';
import messages from '../../../Messages';

/**
 * extend marked's renderer in order to open link in new tab
 */
let renderer = new marked.Renderer();
renderer.link = function() {
    let link = marked.Renderer.prototype.link.apply(this, arguments);
    return link.replace('<a', '<a target=\'_blank\'');
};

marked.setOptions({ renderer });

const CSAwRuleSummary = ({ text, truncate, link, intl }) => {
    const dangerousHtml = (text) => ({ __html: sanitizeHtml(text) });

    return (
        <StackItem>
            <TextContent className="rule-description">
                { truncate && text.length > TRUNCATE_TEXT_THRESHOLD
                    ? (
                        <Truncate
                            length={TRUNCATE_TEXT_THRESHOLD}
                            expandText={intl.formatMessage(messages.readMore)}
                            collapseText={intl.formatMessage(messages.readLess)}
                            text={marked(text)}
                            spaceBetween
                        />
                    ) : (
                        <Stack>
                            <StackItem>
                                <span dangerouslySetInnerHTML={dangerousHtml(marked(text))}/>
                            </StackItem>
                            {link && <StackItem className="rule-link pf-u-mt-sm">
                                {handleCVELink(link, intl.formatMessage(messages.readMore))}
                            </StackItem>}
                        </Stack>
                    )
                }
            </TextContent>
        </StackItem>
    );

};

CSAwRuleSummary.defaultProps = {
    truncate: true,
    link: null
};

CSAwRuleSummary.propTypes = {
    intl: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object
    ]),
    truncate: PropTypes.bool,
    text: PropTypes.string.isRequired,
    link: PropTypes.string
};

export default injectIntl(CSAwRuleSummary);
