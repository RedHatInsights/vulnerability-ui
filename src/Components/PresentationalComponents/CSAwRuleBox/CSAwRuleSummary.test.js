import CSAwRuleSummary from './CSAwRuleSummary'
import { render, screen } from '@testing-library/react';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

describe('CSAW Rule summary', () => {
    const text = `An issue was found in the manner in which certain x86 microprocessors manage thread memory access synchronization. It has been assigned [CVE-2019-11135](https://access.redhat.com/security/cve/CVE-2019-11135). Complete mitigations for this issue require microcode and kernel updates, and disabling hyperthreading where applicable.↵↵An unprivileged user can use this flaw to gain read access to privileged memory that would otherwise be inacessible, including from other virtual machines on the same host system.↵`
    it('should render compenent', () => {
        const { asFragment } = render(
            <IntlProvider locale="en">
                <CSAwRuleSummary text={text} />
            </IntlProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    })

    it('should render truncated text', () => {
        render(
            <IntlProvider locale="en">
                <CSAwRuleSummary truncate text={text} />
            </IntlProvider>
        );

        expect(screen.getAllByText('Read more')).toHaveLength(1);
    })

    it('should render whole text', () => {
        render(
            <IntlProvider locale="en">
                <CSAwRuleSummary truncate={false} text={text} />
            </IntlProvider>
        );

        expect(screen.queryAllByText('Read more')).toHaveLength(0);
    })
});
