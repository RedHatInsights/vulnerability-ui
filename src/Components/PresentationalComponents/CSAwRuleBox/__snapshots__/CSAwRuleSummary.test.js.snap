// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSAW Rule summary should render compenent 1`] = `
<injectIntl(CSAwRuleSummary)
  text="An issue was found in the manner in which certain x86 microprocessors manage thread memory access synchronization. It has been assigned [CVE-2019-11135](https://access.redhat.com/security/cve/CVE-2019-11135). Complete mitigations for this issue require microcode and kernel updates, and disabling hyperthreading where applicable.↵↵An unprivileged user can use this flaw to gain read access to privileged memory that would otherwise be inacessible, including from other virtual machines on the same host system.↵"
>
  <CSAwRuleSummary
    intl={
      {
        "$t": [Function],
        "defaultFormats": {},
        "defaultLocale": "en",
        "defaultRichTextElements": undefined,
        "fallbackOnEmptyString": true,
        "formatDate": [Function],
        "formatDateTimeRange": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatList": [Function],
        "formatListToParts": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": {},
        "formatters": {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "onWarn": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
        "wrapRichTextChunksInFragment": undefined,
      }
    }
    link={null}
    text="An issue was found in the manner in which certain x86 microprocessors manage thread memory access synchronization. It has been assigned [CVE-2019-11135](https://access.redhat.com/security/cve/CVE-2019-11135). Complete mitigations for this issue require microcode and kernel updates, and disabling hyperthreading where applicable.↵↵An unprivileged user can use this flaw to gain read access to privileged memory that would otherwise be inacessible, including from other virtual machines on the same host system.↵"
    truncate={true}
  >
    <StackItem
      link={null}
    >
      <div
        className="pf-l-stack__item"
        link={null}
      >
        <TextContent
          className="rule-description"
        >
          <div
            className="pf-c-content rule-description"
          >
            <Truncate
              collapseText="Read less"
              expandText="Read more"
              length={230}
              spaceBetween={true}
              text="<p>An issue was found in the manner in which certain x86 microprocessors manage thread memory access synchronization. It has been assigned <a target="_blank" rel="noopener noreferrer" href="https://access.redhat.com/security/cve/CVE-2019-11135">CVE-2019-11135</a>. Complete mitigations for this issue require microcode and kernel updates, and disabling hyperthreading where applicable.↵↵An unprivileged user can use this flaw to gain read access to privileged memory that would otherwise be inacessible, including from other virtual machines on the same host system.↵</p>
"
            >
              <Stack
                className="ins-c-truncate is-block"
              >
                <div
                  className="pf-l-stack ins-c-truncate is-block"
                >
                  <StackItem>
                    <div
                      className="pf-l-stack__item"
                    >
                      <span
                        dangerouslySetInnerHTML={
                          {
                            "__html": "<p>An issue was found in the manner in which certain x86 microprocessors manage thread memory access synchronization. It has been assigned </p>",
                          }
                        }
                        widget-type="InsightsTruncateBlock"
                      />
                    </div>
                  </StackItem>
                  <StackItem
                    className="pf-u-mt-sm"
                  >
                    <div
                      className="pf-l-stack__item pf-u-mt-sm"
                    >
                      <Button
                        className="ins-c-expand-button"
                        onClick={[Function]}
                        variant="link"
                      >
                        <ButtonBase
                          className="ins-c-expand-button"
                          innerRef={null}
                          onClick={[Function]}
                          variant="link"
                        >
                          <button
                            aria-disabled={false}
                            aria-label={null}
                            className="pf-c-button pf-m-link ins-c-expand-button"
                            data-ouia-component-id="OUIA-Generated-Button-link-1"
                            data-ouia-component-type="PF4/Button"
                            data-ouia-safe={true}
                            disabled={false}
                            onClick={[Function]}
                            role={null}
                            type="button"
                          >
                            Read more
                          </button>
                        </ButtonBase>
                      </Button>
                    </div>
                  </StackItem>
                </div>
              </Stack>
            </Truncate>
          </div>
        </TextContent>
      </div>
    </StackItem>
  </CSAwRuleSummary>
</injectIntl(CSAwRuleSummary)>
`;
